untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,552 @@
+# A Brief Look at Untyped Lambda Calculus
+
+When talking about Haskell, the term "lambda calculus" often crops up.
+It's a theoretical framework used to define the meaning of computation in many functional languages, such as Haskell, Agda, Idris, etc.
+Understanding lambda calculus can be very helpful when talking about programs written in these languages.
+
+Perhaps surprisingly, understanding lambda calculus can also be helpful in understanding the C++ template metalanguage since it's a functional language in its own right (although it's not, strictly speaking, based on lambda calculus).
```

Comment:
The reader might want to figure out how lambda is helpful in C++ template metalanguage. Are there any references?

Reply:
Well, I can point to (Thielecke. "An introduction to C++ template programming", University of Birmingham, 2016), but that's course notes and are not easy to find.

Beyond that, we have λ-calculus-inspired things in Boost, e.g. [the `lambda` metafunction](https://www.boost.org/doc/libs/1_52_0/libs/mpl/doc/tutorial/the-lambda-metafunction.html). Besides, people have a lot of fun researching the isomorphism between C++ templates and λ-calculus, e.g. (Sinkovics, Porkoláb. "Expressing C++ template metaprograms as lambda expressions." In TFP’09, Jun, pp. 2-4. 2009), (Gozillon, Keir. "Point-Free Templates." In European LLVM Developers Meeting. 2018), and there are a few more, and it is pretty illuminating in its own right (there's a [blog post](https://lettersfromtheoutershell.xyz/2020/10/17/templates-calculus) on this topic too)

Not sure if it makes sense referencing any of those.

Reply:
Yeah, feels like it would be too much.